3_train/src/train.py

```diff
@@ -0,0 +1,465 @@
+    n_static = len(static_features_use)
+
+    # Train/validate split
+    dataset = SequenceDataset(
+        data_array_to_use,
+        n_depths,
+        n_dynamic,
+        n_static
+    )
+    train_subset, valid_subset = split(dataset, 1-valid_frac, seed)
+    return get_data_loaders(train_subset, valid_subset, batch_size)
+
+
+def get_model(
+    n_depths,
+    n_dynamic,
+    n_static,
+    hidden_size,
+    initial_forget_bias,
+    dropout,
+    concat_static
+):
+    """
+    Create LSTM torch model
+    
+    Patterned after https://pytorch.org/tutorials/beginner/nn_tutorial.html#refactor-using-optim
+
+    :param n_depths: Number of depths at which to predict temperatures (number
+        of outputs at each time step)
+    :param n_dynamic: Number of dynamic input features
+    :param n_static: Number of static input features
+    :param hidden_size: Number of elements in hidden state and cell state
+    :param initial_forget_bias: Value of the initial forget gate bias
+    :param dropout: Dropout probability, from 0 to 1 (0 = don't use dropout)
+    :param concat_static: If True, uses standard LSTM. Otherwise, uses EA-LSTM
+    :returns: LSTM torch model
+
+    """
+    if torch.cuda.is_available():
+        device = "cuda" 
+    else:
+        device = "cpu"
+    print(f"Using {device} device")
+
+    model = Model(
+        input_size_dyn=n_dynamic,
+        input_size_stat=n_static,
+        hidden_size=hidden_size,
+        output_size=n_depths,
+        initial_forget_bias=initial_forget_bias,
+        dropout=dropout,
+        concat_static=concat_static
+    ).to(device)
+
+    return model
+
+
+def loss_batch(model, loss_func, x_d, x_s, y, opt=None):
+    """
+    Evaluate loss for one batch of inputs and outputs, and optionally update model
+    parameters by backpropagation
+    
+    Non-finite (e.g. NaN) values in y are ignored.
+    
+    Patterned after https://pytorch.org/tutorials/beginner/nn_tutorial.html#create-fit-and-get-data
+
+    :param model: PyTorch model inheriting nn.Module()
+    :param loss_func: PyTorch loss function, e.g., nn.MSELoss()
+    :param x_d: Dynamic input features
+    :param x_s: Static input features
+    :param y: Outputs (temperatures at multiple depths)
+    :param opt: Optimizer to use to update model parameters. If opt=None, do
+        not update model parameters (Default value = None)
+    :returns: Tuple of loss value and number of finite (non-NaN) output labels
+        in batch
+
+    """
+
+    # model only needs one set of static inputs
+    # They are identical at all time steps
+    pred, h, c = model(x_d, x_s[:, 0, :])
+    # Only compute loss where y is finite
+    loss_idx = torch.isfinite(y)
+    loss = loss_func(pred[loss_idx], y[loss_idx])
+
+    if opt is not None:
+        loss.backward()
+        opt.step()
+        opt.zero_grad()
+
+    # return loss, number of finite values in y
+    return loss.item(), torch.sum(loss_idx).item()
```

Comment:
Do you ever have a batch that is all `nan` y observations? If so, has this been stress tested for that case (i.e. does line 238 error out with an empty input)

One of the more succinct loss functions for handling `nan` 🚀 . I like including that the training operations are in it too 😄 (241-243)

Reply:
Thanks! This loss function is a hybrid of [this PyTorch tutorial](https://pytorch.org/tutorials/beginner/nn_tutorial.html#create-fit-and-get-data) and [Jared Willard's surface temp code](https://github.com/jdwillard19/lake_conus_surface_temp_2021/blob/88334091dec71ae43fe4256603d65045141936b5/src/train/EALSTM_final_model.py#L580-L586). Good point about a batch of all `nan` observations. Every sequence in the training and testing set should include at least one observation. It would be easy to check for this and raise an error, in case of bug.

Reply:
For sure, I was just thinking it would be good to know before you get (e.g.,) multiple hours/days into training. Could use a simply `try/catch` set up to handle the error

Reply:
Yes, `NaN`s sneaking into the loss are the pits! I tested `loss_batch` with all `NaN` labels, and it silently returned `NaN` as the loss value. So, I added a check that raises a `ValueError`. I also added a check to make sure the return value isn't `NaN`.

Reply:
Cool!

I don't know what your batch size or `NaN` percentage is like, but I'm thinking that this would still crash your whole model run if it happens upon a batch of all `NaN` (effectively undo'ing hours/days of work). This is more likely to be the case if you have a small batch size and lots of `NaN` (maybe this isn't the case though and that's why the next part isn't necessary, so feel free to disregard if I'm trying to solve a problem that doesn't exist)

I think a preferable approach would be to skip past a batch if it doesn't have any real observations to train with - e.g., with partial pseudocode

```
for batch_x, batch_y in data_loader:
    if sum(torch.isfinite(batch_y)) == 0:
        continue
    else:
        compute loss
```

This would prevent the loss from being `NaN` and not error out the pipeline

Reply:
I see your point. That would crash the training, potentially after running for a few days. However, in the case of this project, that crash would indicate that there's probably a bug in `2_process` that should be fixed. There shouldn't be a single training (or testing) sequence that consists of all `NaN` observations, much less a batch of them. During `2_process`, only sequences that include at least one finite temperature observation are included in the training and testing sets.

Reply:
I lean toward leaving the code as is. If that error is never raised - great. If the error is reached, the results shouldn't be usable anyway since there's an unexpected problem with the training data that needs to be addressed - something's wrong with the way we're forming the training data in that case.